### PR TITLE
crush,mon: fix weight-set vs crush device classes

### PR DESCRIPTION
**Pull request by liewegas**
- rebuild weight-sets along with shadow hierarchy
- remove old choose_args when we delete the pool

Fixes http://tracker.ceph.com/issues/20939

### CODE REVIEW

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -1702,6 +1702,10 @@ int CrushWrapper::add_bucket(
 
 int CrushWrapper::bucket_add_item(crush_bucket *bucket, int item, int weight)
 {
+  int r = crush_bucket_add_item(crush, bucket, item, weight);
+  if (r < 0) {
+    return r;
+  }
   __u32 new_size = bucket->size + 1;
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: `crush_bucket_add_item` will have done bucket-size adjustment by now, so this is mis-counting here

File: src/crush/CrushWrapper.cc [hunk 2]
```diff
@@ -1721,11 +1725,15 @@ int CrushWrapper::bucket_add_item(crush_bucket *bucket, int item, int weight)
...
+  return 0;
 }
 
 int CrushWrapper::bucket_remove_item(crush_bucket *bucket, int item)
 {
+  int r = crush_bucket_remove_item(crush, bucket, item);
+  if (r < 0) {
+    return r;
+  }
   __u32 new_size = bucket->size - 1;
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: same here

File: src/crush/CrushWrapper.cc [hunk 3]
```diff
@@ -1883,14 +1898,51 @@ int CrushWrapper::device_class_clone(
...
     return res;
+
   // we do not use set_item_name because the name is intentionally invalid
   name_map[*clone] = copy_name;
   if (have_rmaps)
     name_rmap[copy_name] = *clone;
   class_bucket[original_id][device_class] = *clone;
+
+  // set up choose_args for the new bucket. 
+  for (auto w : choose_args) {
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: `for (auto &w : choose_args) `   need & here..

File: src/crush/CrushWrapper.cc [hunk 4]
```diff
@@ -1883,14 +1898,51 @@ int CrushWrapper::device_class_clone(
...
+    crush_choose_arg_map cmap = w.second;
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: need **&** too

File: src/crush/CrushWrapper.cc [hunk 5]
```diff
@@ -1883,14 +1898,51 @@ int CrushWrapper::device_class_clone(
...
+    if (-1-bno >= (int)cmap.size) {
+      unsigned new_size = -1-bno;
```
Comment (xiexingguo) [inline src/crush/CrushWrapper.cc]: since `-1-bno` works as an subscript, I think it instead should be:

`unsigned new_size = -1-bno + 1;`

**Comment by xiexingguo** [comment]
@liewegas Does not compile, need more fixes

**Comment by xiexingguo** [comment]
Two more potential issues:
(1) what if pool-size changed? Do we need to develop a mechanism to sync carg_map with that? 
(2) see https://github.com/ceph/ceph/pull/16672

**Comment by liewegas** [comment]
1. The carg stuff is written to handle weight_set_size != pool_size (in particular < pool_size, which is what 'flat' mode is).  The mon command oversimplifies by only offering flat and positional.  So nothing will break, at least.  I can't decide whether it makes sense to expose all of the underlying functionality or not... at this point, though, what we have now isn't problematic IMO, just incomplete.

**Comment by xiexingguo** [comment]
> The carg stuff is written to handle weight_set_size != pool_size (in particular < pool_size, which is what 'flat' mode is). The mon command oversimplifies by only offering flat and positional. So nothing will break, at least. I can't decide whether it makes sense to expose all of the underlying functionality or not... at this point, though, what we have now isn't problematic IMO, just incomplete.

@liewegas Fine with me.

**Comment by yuriw** [comment]
http://pulpito.ceph.com/yuriw-2017-08-08_22:28:07-rados-wip-yuri-testing4_2017_8_8-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-08-09_14:33:05-rados-wip-yuri-testing4_2017_8_8-distro-basic-smithi/

per @liewegas needs a retest

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-08-12_21:09:40-rados-wip-sage-testing-20170812a-distro-basic-smithi/

**Pull request merged**